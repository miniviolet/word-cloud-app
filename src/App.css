.App {
  text-align: center;
  display: grid;
  grid-template-columns: 2fr 1fr;
  grid-template-rows: 1fr 3fr;
  grid-gap: 25px;
  grid-template-areas: "header header"
  "cloud sidebar";
  font-family: Arial, Helvetica, sans-serif;
}

.header {
  background-color: #282c34;
  color: white;
  grid-area: header;
}

.word-cloud {
  grid-area: cloud;
  display: flex;
  flex-wrap: wrap;
  align-items: center;
  justify-content: center;
}

.sidebar {
  grid-area: sidebar;
  text-align: left;
}

.App-link {
  color: #61dafb;
}

.topic {
  color: #4a4a4a;
  background-color: transparent;
  border: none;
  cursor: pointer;
}

.positive {
  color: #39c65a;
}

.negative {
  color: #e61919;
}

[data-level="6"] {
  font-size: 65px;
  line-height: 70px;
}

[data-level="5"] {
  font-size: 56px;
  line-height: 62px;
}

[data-level="4"] {
  font-size: 45px;
  line-height: 48px;
}

[data-level="3"] {
  font-size: 32px;
  line-height: 36px;
}

[data-level="2"] {
  font-size: 20px;
  line-height: 24px;
}

[data-level="1"] {
  font-size: 16px;
  line-height: 20px;
}

