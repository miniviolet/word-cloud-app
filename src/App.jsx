import './App.css';
import { useEffect, useState } from 'react';
import Topics from './assets/topics'
import Topic from './components/Topic';
import Metadata from './components/Metadata';

function App() {
  const topics = Topics.topics
  const [currentIndex, setCurrentIndex] = useState(0)
  const [currentWord, setCurrentWord] = useState(topics[0])
  const [randomOrderList, setRandomOrderList] = useState(Array.from({length: topics.length}, () => Math.floor(Math.random() * topics.length)))

  useEffect(() => {
    setCurrentWord(topics[currentIndex])
  }, [currentIndex, topics])

  useEffect(() => {
    setRandomOrderList(Array.from({length: topics.length}, () => Math.floor(Math.random() * topics.length)))
  }, [topics])

  const scores = topics.map(({ volume }) => volume)
  scores.sort(function(a, b) {
    return a - b;
  });
  const range = (scores[scores.length - 1] - scores[0]) / 6

  const whichRange = (x) => {
    let rangeScore = 0
    for (let index = 0; index < 6; index++) {
      let bottomRange = scores[0] + (range * index)
      let topRange = bottomRange + range
      if(bottomRange < x && x <= topRange) rangeScore = index + 1
    }
    return rangeScore
  }

  return (
    <div className="App">
      <header className='header'>
        <h1>My Topics Challenge</h1>
      </header>
      <section className='word-cloud'>
        {
          topics.map((topic, idx) => {
            const { id, label, sentimentScore, volume } = topic
            const level = whichRange(volume)
           return <Topic key={id} {...{label, onClick: () => setCurrentIndex(idx), level, randomOrder: randomOrderList[idx], sentimentScore}} />
          })
        }
      </section>
      <section className='sidebar'>
        <Metadata {...{label: currentWord.label, sentiment: currentWord.sentiment, volume: currentWord.volume}} />
      </section>
    </div>
  );
}

export default App;
