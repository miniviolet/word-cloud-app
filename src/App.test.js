import {fireEvent, render, screen } from '@testing-library/react';
import App from './App';
import Topics from './assets/topics'

describe('Word cloud app', () => {
  const topics = Topics.topics

  test('renders Cloud word',() => {
    const container = render(<App />);
    expect(container).toMatchSnapshot()
    expect(screen.getAllByRole('button').length).toEqual(topics.length)
  })

  test('renders Metadata', () => {
    render(<App />);
    expect(screen.getByTestId('metadata-header')).toBeInTheDocument()
    expect(screen.getByTestId('metadata-header')).toHaveTextContent(topics[0].label)
    expect(screen.getByTestId('positive-mention')).toHaveTextContent(topics[0].sentiment.positive)
    expect(screen.getByTestId('neutral-mention')).toHaveTextContent(topics[0].sentiment.neutral)
    expect(screen.getByTestId('negative-mention')).toHaveTextContent(topics[0].sentiment.negative)
  })

  test('updates Metadata on click', () => {
    render(<App />);
    const buttonClicked = screen.getAllByRole('button')[1]
    fireEvent.click(buttonClicked)
    expect(screen.getByTestId('metadata-header')).toHaveTextContent(topics[1].label)
    expect(screen.getByTestId('positive-mention')).toHaveTextContent(topics[1].sentiment.positive)
    expect(screen.getByTestId('neutral-mention')).toHaveTextContent(topics[1].sentiment.neutral)
    expect(screen.getByTestId('negative-mention')).toHaveTextContent(0)
  })
});
