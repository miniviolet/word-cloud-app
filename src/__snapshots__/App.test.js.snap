// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Word cloud app renders Cloud word 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="App"
      >
        <header
          class="header"
        >
          <h1>
            My Topics Challenge
          </h1>
        </header>
        <section
          class="word-cloud"
        >
          <button
            class="topic positive"
            data-level="6"
            style="order: 18;"
          >
            Berlin
          </button>
          <button
            class="topic "
            data-level="2"
            style="order: 20;"
          >
            DJ
          </button>
          <button
            class="topic "
            data-level="1"
            style="order: 9;"
          >
            Ostgut Ton
          </button>
          <button
            class="topic negative"
            data-level="2"
            style="order: 23;"
          >
            Hammered
          </button>
          <button
            class="topic positive"
            data-level="1"
            style="order: 8;"
          >
            Code
          </button>
          <button
            class="topic "
            data-level="1"
            style="order: 27;"
          >
            Quantified Drunk
          </button>
          <button
            class="topic positive"
            data-level="1"
            style="order: 11;"
          >
            Berghain resident
          </button>
          <button
            class="topic "
            data-level="1"
            style="order: 24;"
          >
            San Soda's Panorama Bar
          </button>
          <button
            class="topic positive"
            data-level="1"
            style="order: 9;"
          >
            Germany
          </button>
          <button
            class="topic positive"
            data-level="1"
            style="order: 11;"
          >
            Amsterdam
          </button>
          <button
            class="topic "
            data-level="1"
            style="order: 0;"
          >
            Kantine am Berghain
          </button>
          <button
            class="topic positive"
            data-level="1"
            style="order: 9;"
          >
            London
          </button>
          <button
            class="topic "
            data-level="1"
            style="order: 19;"
          >
            UK
          </button>
          <button
            class="topic positive"
            data-level="1"
            style="order: 21;"
          >
            Marcel Dettmann
          </button>
          <button
            class="topic "
            data-level="1"
            style="order: 17;"
          >
            Disco
          </button>
          <button
            class="topic "
            data-level="1"
            style="order: 8;"
          >
            Barcelona
          </button>
          <button
            class="topic positive"
            data-level="1"
            style="order: 18;"
          >
            Watergate
          </button>
          <button
            class="topic positive"
            data-level="1"
            style="order: 14;"
          >
            debut LP
          </button>
          <button
            class="topic positive"
            data-level="1"
            style="order: 22;"
          >
            Patrick Gräser
          </button>
          <button
            class="topic positive"
            data-level="1"
            style="order: 26;"
          >
            Panorama Bar in Berlin
          </button>
          <button
            class="topic positive"
            data-level="1"
            style="order: 0;"
          >
            legendary nightclub
          </button>
          <button
            class="topic "
            data-level="1"
            style="order: 18;"
          >
            Ben Klock
          </button>
          <button
            class="topic "
            data-level="1"
            style="order: 15;"
          >
            Mixes
          </button>
          <button
            class="topic "
            data-level="1"
            style="order: 26;"
          >
            Panorama Bar Music
          </button>
          <button
            class="topic "
            data-level="1"
            style="order: 21;"
          >
            Terrace Sundae
          </button>
          <button
            class="topic "
            data-level="1"
            style="order: 14;"
          >
            Jun
          </button>
          <button
            class="topic positive"
            data-level="1"
            style="order: 3;"
          >
            Live set
          </button>
          <button
            class="topic "
            data-level="1"
            style="order: 5;"
          >
            dance music
          </button>
          <button
            class="topic "
            data-level="0"
            style="order: 24;"
          >
            club culture
          </button>
          <button
            class="topic "
            data-level="0"
            style="order: 20;"
          >
            D/B Presents
          </button>
        </section>
        <section
          class="sidebar"
        >
          <p
            data-testid="metadata-header"
          >
            Information on topic "
            Berlin
            "
          </p>
          <p>
            Total mentions: 
            165
          </p>
          <p>
            Positive mentions: 
            <span
              class="positive"
              data-testid="positive-mention"
            >
              29
            </span>
          </p>
          <p>
            Neutral mentions: 
            <span
              data-testid="neutral-mention"
            >
              133
            </span>
          </p>
          <p>
            Negative mentions: 
            <span
              class="negative"
              data-testid="negative-mention"
            >
              3
            </span>
          </p>
        </section>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="App"
    >
      <header
        class="header"
      >
        <h1>
          My Topics Challenge
        </h1>
      </header>
      <section
        class="word-cloud"
      >
        <button
          class="topic positive"
          data-level="6"
          style="order: 18;"
        >
          Berlin
        </button>
        <button
          class="topic "
          data-level="2"
          style="order: 20;"
        >
          DJ
        </button>
        <button
          class="topic "
          data-level="1"
          style="order: 9;"
        >
          Ostgut Ton
        </button>
        <button
          class="topic negative"
          data-level="2"
          style="order: 23;"
        >
          Hammered
        </button>
        <button
          class="topic positive"
          data-level="1"
          style="order: 8;"
        >
          Code
        </button>
        <button
          class="topic "
          data-level="1"
          style="order: 27;"
        >
          Quantified Drunk
        </button>
        <button
          class="topic positive"
          data-level="1"
          style="order: 11;"
        >
          Berghain resident
        </button>
        <button
          class="topic "
          data-level="1"
          style="order: 24;"
        >
          San Soda's Panorama Bar
        </button>
        <button
          class="topic positive"
          data-level="1"
          style="order: 9;"
        >
          Germany
        </button>
        <button
          class="topic positive"
          data-level="1"
          style="order: 11;"
        >
          Amsterdam
        </button>
        <button
          class="topic "
          data-level="1"
          style="order: 0;"
        >
          Kantine am Berghain
        </button>
        <button
          class="topic positive"
          data-level="1"
          style="order: 9;"
        >
          London
        </button>
        <button
          class="topic "
          data-level="1"
          style="order: 19;"
        >
          UK
        </button>
        <button
          class="topic positive"
          data-level="1"
          style="order: 21;"
        >
          Marcel Dettmann
        </button>
        <button
          class="topic "
          data-level="1"
          style="order: 17;"
        >
          Disco
        </button>
        <button
          class="topic "
          data-level="1"
          style="order: 8;"
        >
          Barcelona
        </button>
        <button
          class="topic positive"
          data-level="1"
          style="order: 18;"
        >
          Watergate
        </button>
        <button
          class="topic positive"
          data-level="1"
          style="order: 14;"
        >
          debut LP
        </button>
        <button
          class="topic positive"
          data-level="1"
          style="order: 22;"
        >
          Patrick Gräser
        </button>
        <button
          class="topic positive"
          data-level="1"
          style="order: 26;"
        >
          Panorama Bar in Berlin
        </button>
        <button
          class="topic positive"
          data-level="1"
          style="order: 0;"
        >
          legendary nightclub
        </button>
        <button
          class="topic "
          data-level="1"
          style="order: 18;"
        >
          Ben Klock
        </button>
        <button
          class="topic "
          data-level="1"
          style="order: 15;"
        >
          Mixes
        </button>
        <button
          class="topic "
          data-level="1"
          style="order: 26;"
        >
          Panorama Bar Music
        </button>
        <button
          class="topic "
          data-level="1"
          style="order: 21;"
        >
          Terrace Sundae
        </button>
        <button
          class="topic "
          data-level="1"
          style="order: 14;"
        >
          Jun
        </button>
        <button
          class="topic positive"
          data-level="1"
          style="order: 3;"
        >
          Live set
        </button>
        <button
          class="topic "
          data-level="1"
          style="order: 5;"
        >
          dance music
        </button>
        <button
          class="topic "
          data-level="0"
          style="order: 24;"
        >
          club culture
        </button>
        <button
          class="topic "
          data-level="0"
          style="order: 20;"
        >
          D/B Presents
        </button>
      </section>
      <section
        class="sidebar"
      >
        <p
          data-testid="metadata-header"
        >
          Information on topic "
          Berlin
          "
        </p>
        <p>
          Total mentions: 
          165
        </p>
        <p>
          Positive mentions: 
          <span
            class="positive"
            data-testid="positive-mention"
          >
            29
          </span>
        </p>
        <p>
          Neutral mentions: 
          <span
            data-testid="neutral-mention"
          >
            133
          </span>
        </p>
        <p>
          Negative mentions: 
          <span
            class="negative"
            data-testid="negative-mention"
          >
            3
          </span>
        </p>
      </section>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
